libcst/_exceptions.py

```diff
@@ -4,77 +4,225 @@
+
+
+# pyre-fixme[2]: 'Any' type isn't pyre-strict.
+def _parser_syntax_error_unpickle(kwargs: Any) -> "ParserSyntaxError":
+    return ParserSyntaxError(**kwargs)
+
+
+@final
 class ParserSyntaxError(Exception):
     """
-    Contains error information about the parser tree.
+    Contains an error encountered while trying to parse a piece of source code. This
+    exception shouldn't be constructed directly by the user, but instead may be raised
+    by calls to :func:`parse_module`, :func:`parse_expression`, or
+    :func:`parse_statement`.
+
+    This does not inherit from :class:`SyntaxError` because Python's may raise a
+    :class:`SyntaxError` for any number of reasons, potentially leading to unintended
+    behavior.
     """
 
+    #: A human-readable explanation of the syntax error without information about where
+    #: the error occurred.
+    #:
+    #: For a human-readable explanation of the error alongside information about where
+    #: it occurred, use :meth:`__str__` (via ``str(ex)``) instead.
+    message: str
+
+    # An internal value used to compute `editor_column` and to pretty-print where the
+    # syntax error occurred in the code.
+    _lines: Sequence[str]
+
+    # These are internal values for the public raw_line/raw_column properties. They may
+    # get backfilled later.
+    _raw_line: int
+    _raw_column: int
+
     def __init__(
         self,
         message: str,
-        encountered: Optional[str],  # None means EOF
-        expected: Optional[Iterable[str]],  # None means EOF
-        pos: Tuple[int, int],  # (one-indexed line, zero-indexed column)
-        lines: Sequence[str],  # source code, used to generate a human-readable output
+        *,
+        lines: Sequence[str],
+        raw_line: int = -1,
+        raw_column: int = -1,
     ) -> None:
+        super(ParserSyntaxError, self).__init__(message)
         self.message = message
-        self.encountered = encountered
-        self.expected = expected
-        self.pos = pos
-        self.lines = lines
-        # Make pickling this error work
-        super(ParserSyntaxError, self).__init__(
-            message, encountered, expected, pos, lines
+        self._lines = lines
+        self._raw_line = raw_line
+        self._raw_column = raw_column
+
+    def __reduce__(
+        self
+    ) -> Tuple[Callable[..., "ParserSyntaxError"], Tuple[object, ...]]:
+        return (
+            _parser_syntax_error_unpickle,
+            (
+                {
+                    "message": self.message,
+                    "lines": self._lines,
+                    "raw_line": self._raw_line,
+                    "raw_column": self._raw_column,
+                },
+            ),
         )
 
     def __str__(self) -> str:
         """
-        A human-readable error message of where the syntax error is in their code.
+        A multi-line human-readable error message of where the syntax error is in their
+        code.
```

Comment:
Can we add an example error message in this comment?

Reply:
:+1: Will do.